src/email.rs

```diff
@@ -8,126 +9,175 @@ use lettre::transport::smtp::SmtpTransport;
+    }
 
-fn send_email(recipient: &str, subject: &str, body: String) -> AppResult<()> {
-    let mailgun_config = init_config_vars();
-    let email = build_email(recipient, subject, body, &mailgun_config)?;
-
-    match mailgun_config {
-        Some(mailgun_config) => {
-            let transport = SmtpTransport::relay(&mailgun_config.smtp_server)
-                .unwrap()
-                .credentials(Credentials::new(
-                    mailgun_config.smtp_login,
-                    mailgun_config.smtp_password,
-                ))
-                .authentication(vec![Mechanism::Plain])
-                .build();
-
-            let result = transport.send(&email);
-            result.map_err(|_| server_error("Error in sending email"))?;
+    fn send(&self, recipient: &str, subject: &str, body: &str) -> AppResult<()> {
+        let email = Message::builder()
+            .to(recipient.parse()?)
+            .from(self.sender_address().parse()?)
+            .subject(subject)
+            .body(body.to_string())?;
+
+        match &self.backend {
+            EmailBackend::Smtp {
+                server,
+                login,
+                password,
+            } => {
+                SmtpTransport::relay(&server)?
+                    .credentials(Credentials::new(login.clone(), password.clone()))
+                    .authentication(vec![Mechanism::Plain])
+                    .build()
+                    .send(&email)
+                    .map_err(|_| server_error("Error in sending email"))?;
+            }
+            EmailBackend::FileSystem { path } => {
+                FileTransport::new(&path)
+                    .send(&email)
+                    .map_err(|_| server_error("Email file could not be generated"))?;
+            }
+            EmailBackend::Memory { mails } => mails.lock().unwrap().push(StoredEmail {
+                to: recipient.into(),
+                subject: subject.into(),
+                body: body.into(),
+            }),
         }
-        None => {
-            let sender = FileTransport::new(Path::new("/tmp"));
-            let result = sender.send(&email);
-            result.map_err(|_| server_error("Email file could not be generated"))?;
+
+        Ok(())
+    }
+
+    fn sender_address(&self) -> &str {
+        match &self.backend {
+            EmailBackend::Smtp { login, .. } => login,
+            EmailBackend::FileSystem { .. } => "test@localhost",
+            EmailBackend::Memory { .. } => "test@localhost",
+        }
+    }
+}
+
+enum EmailBackend {
+    /// Backend used in production to send mails using SMTP.
+    Smtp {
+        server: String,
+        login: String,
+        password: String,
+    },
+    /// Backend used locally during development, will store the emails in the provided directory.
+    FileSystem { path: PathBuf },
+    /// Backend used during tests, will keep messages in memory to allow tests to retrieve them.
+    Memory { mails: Mutex<Vec<StoredEmail>> },
+}
```

Comment:
what we're building here is essentially an interface with three backing implementations. I'm curious why you chose to implement that with an `enum`. would something like `Box<dyn EmailTrait>` work too?

Reply:
With few variants I personally prefer `enum`s rather than a `Box<dyn T>`, but either could work just fine.